diff --git a/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CDMALTEPhone.java b/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CDMALTEPhone.java
index ea943d8..dad5a7b 100644
--- a/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CDMALTEPhone.java
+++ b/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CDMALTEPhone.java
@@ -233,11 +233,14 @@ public class CDMALTEPhone extends CDMAPhone {
         return false;
     }
 
-    // return IMSI from USIM as subscriber ID.
-    @Override
-    public String getSubscriberId() {
-        return (mSimRecords != null) ? mSimRecords.getIMSI() : "";
-    }
+   @Override
+   public String getSubscriberId() {
+       if ((super.getSubscriberId()) != null) {
+           return super.getSubscriberId();
+       } else {
+           return (mSimRecords != null) ? mSimRecords.getIMSI() : "";
+       }
+   }
 
     // return GID1 from USIM
     @Override
diff --git a/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CDMAPhone.java b/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CDMAPhone.java
index 5a683c1..00b7b9e 100644
--- a/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CDMAPhone.java
+++ b/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CDMAPhone.java
@@ -654,7 +654,15 @@ public class CDMAPhone extends PhoneBase {
 
     @Override
     public String getSubscriberId() {
-        return mSST.getImsi();
+        if (mCdmaSubscriptionSource == CDMA_SUBSCRIPTION_NV) {
+            return mSST.getNvImsi();
+        } else {
+            IccRecords r = mIccRecords.get();
+            if (r != null) {
+                return r.getIMSI();
+            }
+        }
+        return null;
     }
 
     @Override
diff --git a/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CdmaServiceStateTracker.java b/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CdmaServiceStateTracker.java
index dcc5a01..2923de4 100644
--- a/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CdmaServiceStateTracker.java
+++ b/frameworks/opt/telephony/src/java/com/android/internal/telephony/cdma/CdmaServiceStateTracker.java
@@ -434,17 +434,6 @@ public class CdmaServiceStateTracker extends ServiceStateTracker {
                     mIsMinInfoReady = true;
 
                     updateOtaspState();
-                    if (!mIsSubscriptionFromRuim && mIccRecords != null) {
-                        if (DBG) {
-                            log("GET_CDMA_SUBSCRIPTION set imsi in mIccRecords");
-                        }
-                        mIccRecords.setImsi(getImsi());
-                    } else {
-                        if (DBG) {
-                            log("GET_CDMA_SUBSCRIPTION either mIccRecords is null  or NV type device" +
-                                    " - not setting Imsi in mIccRecords");
-                        }
-                    }
                 } else {
                     if (DBG) {
                         log("GET_CDMA_SUBSCRIPTION: error parsing cdmaSubscription params num="
@@ -1922,12 +1911,11 @@ public class CdmaServiceStateTracker extends ServiceStateTracker {
     }
 
     /**
-     * Returns IMSI as MCC + MNC + MIN
+     * Returns IMSI from NV in the format MCC + MNC + MIN
      */
-    String getImsi() {
-        // TODO: When RUIM is enabled, IMSI will come from RUIM not build-time props.
+    public String getNvImsi() {
         String operatorNumeric = getSystemProperty(
-                TelephonyProperties.PROPERTY_ICC_OPERATOR_NUMERIC, "");
+                CDMAPhone.PROPERTY_CDMA_HOME_OPERATOR_NUMERIC, "");
 
         if (!TextUtils.isEmpty(operatorNumeric) && getCdmaMin() != null) {
             return (operatorNumeric + getCdmaMin());
diff --git a/frameworks/opt/telephony/src/java/com/android/internal/telephony/dataconnection/DcTracker.java b/frameworks/opt/telephony/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
index ddc3b32..a1a9152 100755
--- a/frameworks/opt/telephony/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
+++ b/frameworks/opt/telephony/src/java/com/android/internal/telephony/dataconnection/DcTracker.java
@@ -908,6 +908,7 @@ public final class DcTracker extends DcTrackerBase {
     }
 
     private boolean trySetupData(ApnContext apnContext) {
+        boolean retValue = false;
         if (DBG) {
             log("trySetupData for type:" + apnContext.getApnType() +
                     " due to " + apnContext.getReason() + " apnContext=" + apnContext);
@@ -950,10 +951,13 @@ public final class DcTracker extends DcTrackerBase {
                 ArrayList<ApnSetting> waitingApns = buildWaitingApns(apnContext.getApnType(),
                         radioTech);
                 if (waitingApns.isEmpty()) {
-                    notifyNoData(DcFailCause.MISSING_UNKNOWN_APN, apnContext);
                     notifyOffApnsOfAvailability(apnContext.getReason());
-                    if (DBG) log("trySetupData: X No APN found retValue=false");
-                    return false;
+                    retValue = setupData(apnContext, radioTech);
+                    if(!retValue) {
+                        notifyNoData(DcFailCause.MISSING_UNKNOWN_APN, apnContext);
+                    }
+                    notifyOffApnsOfAvailability(apnContext.getReason());
+                    return retValue;
                 } else {
                     apnContext.setWaitingApns(waitingApns);
                     if (DBG) {
@@ -972,7 +976,9 @@ public final class DcTracker extends DcTrackerBase {
                             + apnListToString(apnContext.getWaitingApns()));
                 }
             }
-            boolean retValue = setupData(apnContext, radioTech);
+
+            retValue = setupData(apnContext, radioTech);
+
             notifyOffApnsOfAvailability(apnContext.getReason());
 
             if (DBG) log("trySetupData: X retValue=" + retValue);
@@ -1378,8 +1384,38 @@ public final class DcTracker extends DcTrackerBase {
 
         apnSetting = apnContext.getNextWaitingApn();
         if (apnSetting == null) {
-            if (DBG) log("setupData: return for no apn found!");
-            return false;
+            if(PhoneConstants.PHONE_TYPE_CDMA==mPhone.getPhoneType()) {
+                String[] mDunApnTypes = { PhoneConstants.APN_TYPE_DUN };
+                final int mDefaultApnId = DctConstants.APN_DEFAULT_ID;
+                final String[] mDefaultApnTypes = {
+                    PhoneConstants.APN_TYPE_DEFAULT,
+                    PhoneConstants.APN_TYPE_MMS,
+                    PhoneConstants.APN_TYPE_SUPL,
+                    PhoneConstants.APN_TYPE_HIPRI,
+                    PhoneConstants.APN_TYPE_FOTA,
+                    PhoneConstants.APN_TYPE_IMS,
+                    PhoneConstants.APN_TYPE_CBS};
+
+
+                String[] types;
+                int apnId;
+                if (mRequestedApnType.equals(PhoneConstants.APN_TYPE_DUN)) {
+                    types = mDunApnTypes;
+                    apnId = DctConstants.APN_DUN_ID;
+                } else {
+                    types = mDefaultApnTypes;
+                    apnId = mDefaultApnId;
+                }
+                apnSetting = new ApnSetting(apnId, getOperatorNumeric(), null, null,
+                                null, null, null, null, null, null, null,
+                    RILConstants.SETUP_DATA_AUTH_PAP_CHAP, types,
+                    PROPERTY_CDMA_IPPROTOCOL, PROPERTY_CDMA_ROAMING_IPPROTOCOL, true, 0,
+                    0, false, 0, 0, 0, PhoneConstants.UNSET_MTU, "", "");
+                if (DBG) log("setupData: CDMA detected and apnSetting == null, use stubbed CDMA APN setting= " + apnSetting);
+            } else {
+                if (DBG) log("setupData: return for no apn found!");
+                return false;
+            }
         }
 
         int profileId = apnSetting.profileId;
diff --git a/frameworks/opt/telephony/src/java/com/android/internal/telephony/imsphone/ImsPhone.java b/frameworks/opt/telephony/src/java/com/android/internal/telephony/imsphone/ImsPhone.java
index e41de45..2a6fc2a 100644
--- a/frameworks/opt/telephony/src/java/com/android/internal/telephony/imsphone/ImsPhone.java
+++ b/frameworks/opt/telephony/src/java/com/android/internal/telephony/imsphone/ImsPhone.java
@@ -1201,6 +1201,7 @@ public class ImsPhone extends ImsPhoneBase {
             for (int i = 0, s = infos.length; i < s; i++) {
                 if (infos[i].mCondition == ImsUtInterface.CDIV_CF_UNCONDITIONAL) {
                     if (r != null) {
+                        setCallForwardingPreference(infos[i].mStatus == 1);
                         r.setVoiceCallForwardingFlag(1, (infos[i].mStatus == 1),
                             infos[i].mNumber);
                     }
@@ -1270,9 +1271,11 @@ public class ImsPhone extends ImsPhoneBase {
                 IccRecords r = getIccRecords();
                 Cf cf = (Cf) ar.userObj;
                 if (cf.mIsCfu && ar.exception == null && r != null) {
+                    setCallForwardingPreference(msg.arg1 == 1);
                     r.setVoiceCallForwardingFlag(1, msg.arg1 == 1, cf.mSetCfNumber);
                 }
                 sendResponse(cf.mOnComplete, null, ar.exception);
+                updateCallForwardStatus();
                 break;
 
             case EVENT_GET_CALL_FORWARD_DONE:
@@ -1281,6 +1284,7 @@ public class ImsPhone extends ImsPhoneBase {
                     cfInfos = handleCfQueryResult((ImsCallForwardInfo[])ar.result);
                 }
                 sendResponse((Message) ar.userObj, cfInfos, ar.exception);
+                updateCallForwardStatus();
                 break;
 
              case EVENT_SET_CALL_FORWARD_TIMER_DONE:
diff --git a/frameworks/opt/telephony/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java b/frameworks/opt/telephony/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java
index 8ada078..9222878 100644
--- a/frameworks/opt/telephony/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java
+++ b/frameworks/opt/telephony/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java
@@ -1604,6 +1604,7 @@ public final class ImsPhoneCallTracker extends CallTracker {
             // Make mIsSrvccCompleted flag to true after SRVCC complete.
             // After SRVCC complete sometimes SRV_STATUS_UPDATE come late.
             mIsSrvccCompleted = true;
+            mState = PhoneConstants.State.IDLE;
         }
     }
 
diff --git a/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/IccCardProxy.java b/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/IccCardProxy.java
index c4c55c6..17bae02 100755
--- a/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/IccCardProxy.java
+++ b/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/IccCardProxy.java
@@ -127,19 +127,6 @@ public class IccCardProxy extends Handler implements IccCard {
     private boolean mIsCardStatusAvailable = false;
     private PersoSubState mPersoSubState = PersoSubState.PERSOSUBSTATE_UNKNOWN;
 
-    // Sim State events may be broadcasted before the siminfo table update has been
-    // completed. Due to this such events may be broadcasted with dummy subId for a
-    // particular slotId. Therefore, setExternalState once the siminfo table has been updated.
-    // For example, if the UI receives the sim state broadcast with the state as pin locked
-    // with dummy subId, the pin lock screen will not be displayed.
-    private BroadcastReceiver mReceiver = new BroadcastReceiver() {
-            @Override
-            public void onReceive(Context context, Intent intent) {
-                if (TelephonyIntents.ACTION_SUBINFO_RECORD_UPDATED.equals(intent.getAction())) {
-                    setExternalState(mExternalState, true);
-                }
-            }};
-
     public IccCardProxy(Context context, CommandsInterface ci, int phoneId) {
         if (DBG) log("ctor: ci=" + ci + " phoneId=" + phoneId);
         mContext = context;
@@ -151,10 +138,6 @@ public class IccCardProxy extends Handler implements IccCard {
         mUiccController.registerForIccChanged(this, EVENT_ICC_CHANGED, null);
         ci.registerForOn(this,EVENT_RADIO_ON, null);
         ci.registerForOffOrNotAvailable(this, EVENT_RADIO_OFF_OR_UNAVAILABLE, null);
-        IntentFilter filter =
-                new IntentFilter(TelephonyIntents.ACTION_SUBINFO_RECORD_UPDATED);
-        mContext.registerReceiver(mReceiver, filter);
-
         resetProperties();
         setExternalState(State.NOT_READY, false);
     }
diff --git a/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/IccRecords.java b/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/IccRecords.java
index 50aedf6..d4dbae5 100644
--- a/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/IccRecords.java
+++ b/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/IccRecords.java
@@ -332,16 +332,7 @@ public abstract class IccRecords extends Handler implements IccConstants {
      * @return null if SIM is not yet ready or unavailable
      */
     public String getIMSI() {
-        return null;
-    }
-
-    /**
-     * Imsi could be set by ServiceStateTrackers in case of cdma
-     * @param imsi
-     */
-    public void setImsi(String imsi) {
-        mImsi = imsi;
-        mImsiReadyRegistrants.notifyRegistrants();
+        return mImsi;
     }
 
     /**
diff --git a/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/RuimRecords.java b/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/RuimRecords.java
index b7edcec..030e419 100644
--- a/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/RuimRecords.java
+++ b/frameworks/opt/telephony/src/java/com/android/internal/telephony/uicc/RuimRecords.java
@@ -181,11 +181,6 @@ public final class RuimRecords extends IccRecords {
         mRecordsRequested = false;
     }
 
-    @Override
-    public String getIMSI() {
-        return mImsi;
-    }
-
     public String getMdnNumber() {
         return mMyMobileNumber;
     }
